test/test_rtc.cpp

```diff
@@ -0,0 +1,249 @@
+#include <libtorrent/torrent.hpp>
+#include <libtorrent/magnet_uri.hpp>
+
+#include "test.hpp"
+#include "test_utils.hpp"
+#include "setup_transfer.hpp"
+#include "settings.hpp"
+#include "session_mock.hpp"
+
+#if TORRENT_USE_RTC
+
+#include <boost/asio.hpp>
+
+#include <iostream>
+#include <chrono>
+#include <numeric>
+#include <algorithm>
+#include <random>
+
+#include <cstdio>
+#include <cstdarg>
+
+using namespace std::placeholders;
+using namespace std::chrono_literals;
+using namespace lt;
+
+using aux::rtc_signaling;
+using aux::rtc_stream;
+using aux::rtc_stream_init;
+using aux::rtc_offer;
+using aux::rtc_answer;
+
+namespace {
+
+boost::asio::io_context io_context;
+
+session_mock ses1(io_context);
+torrent tor1(ses1, false, parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"));
+
+session_mock ses2(io_context);
+torrent tor2(ses2, false, parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"));
+
+bool success = false;
+
+void run_test() {
+	success = false;
+
+	auto begin_time = clock_type::now();
+	auto end_time = begin_time + 60s;
+	do {
+		io_context.run_one_until(end_time);
+		io_context.restart();
+	}
+	while (!success && clock_type::now() < end_time);
+
+	ses1.print_alerts(begin_time);
+	ses1.print_alerts(begin_time);
+
+	TEST_CHECK(success);
+}
+
+void test_offers()
+{
+	std::shared_ptr<rtc_signaling> sig;
+
+	const int requested_offers_count = 10;
+
+	auto offers_handler = [&](error_code const& ec, std::vector<rtc_offer> offers) {
+		TEST_CHECK(!ec);
+		TEST_EQUAL(int(offers.size()), requested_offers_count);
+		success = true;
+	};
+
+	auto handler = [&](peer_id const&, rtc_stream_init) {};
+
+	sig = std::make_shared<rtc_signaling>(io_context, &tor1, handler);
+	sig->generate_offers(requested_offers_count, offers_handler);
+
+	run_test();
+
+	sig->close();
+}
+
+void test_connectivity()
+{
+	std::shared_ptr<rtc_signaling> sig1, sig2;
+	bool endpoint1 = false;
+	bool endpoint2 = false;
+
+	auto answer_callback = [&](peer_id const&, rtc_answer const& answer) {
+		sig1->process_answer(answer);
+	};
+
+	auto offers_handler = [&](error_code const& ec, std::vector<rtc_offer> offers) {
+		TEST_CHECK(!ec);
+		TEST_EQUAL(int(offers.size()), 1);
+
+		rtc_offer offer = offers[0];
+		offer.answer_callback = answer_callback;
+		sig2->process_offer(offer);
+	};
+
+	auto handler1 = [&](peer_id const&, rtc_stream_init init) {
+		TEST_CHECK(init.peer_connection);
+		TEST_CHECK(init.data_channel);
+		endpoint1 = true;
+		success = endpoint2;
+	};
+
+	auto handler2 = [&](peer_id const&, rtc_stream_init init) {
+		TEST_CHECK(init.peer_connection);
+		TEST_CHECK(init.data_channel);
+		endpoint2 = true;
+		success = endpoint1;
+	};
+
+	sig1 = std::make_shared<rtc_signaling>(io_context, &tor1, handler1);
+	sig2 = std::make_shared<rtc_signaling>(io_context, &tor2, handler2);
+
+	sig1->generate_offers(1, offers_handler);
+
+	run_test();
+
+	sig1->close();
+	sig2->close();
+}
+
+void test_stream()
```

Comment:
this test also fails for me on linux.
```
***** test_rtc.cpp:92 "TEST_ERROR: check failed: "success"" *****


***** test_rtc.cpp:227 "TEST_ERROR: check failed: "written"" *****


***** test_rtc.cpp:229 "TEST_ERROR: check failed: "stream1"" *****


***** test_rtc.cpp:230 "TEST_ERROR: check failed: "stream2"" *****



[test_rtc.cpp.signaling_stream      ] 4 FAILURES
```

I wonder if it has to do with the python websocket. I have python 3.8. But I don't see any error messages (or any output at all actually) from python.

Reply:
The websocket tracker (the one using the python websocket server) is actually an independant one in `test_tracker.cpp`. I wonder why this one fails on your machine though. I've seen a similar issue where it fails to connect locally, which is what the test tries to do, however on macOS. I'm investigating.

Reply:
could you make it print more output?

Reply:
I fixed the local connection failure (it seemed like a firewall issue on macOS, somehow a local connection with an address different from localhost causes a permission error) and added more output, could you retry please?